---
title: "AFP_LongRead_Notebook"
author: "Sam Bogan and Nathan Surendran"
output:
  html_document: default
  pdf_document: default
---

#Summary

This is an R markdown lab notebook for all research, wet lab work, and analyses related to a study on type AFP III evolution revealed by long read sequencing and genome assembly in Zoarcoid fishes. The 11 Zoarcoid species included in the study and Scorpaeiformes outgroups (sticklebacks) are indexed

File registry
1. Metadata
  + LongRead_SpeciesIDs.csv (index of species names and genome IDs)

\br

Below is a quick visual representation of the species' relatedness using the rfishtree time-calibrated phylogeny

```{r setup, include=FALSE}

knitr::opts_knit$set(root.dir = '~/Documents/GitHub/Long_AFP/')

# Load packages for plotting quick species tree
library(tidyverse)
library(fishtree)
library(ggtree)
library(MatrixModels)
library(rfishbase)
library(phytools)

```

What genomes are we working with?

```{r}

# Read in list of species
Species_df <- read.csv("LongRead_SpeciesIDs.csv")

# Print list to show which genomes came from public source vs. Kelley Lab
Species_df

```

Extract and plot species phylogeny with important metadata

```{r, results = FALSE}

# Extract eelpout phylogeny
species_phy <- fishtree_phylogeny(species = Species_df$Species_ID, type=c("chronogram"))

# How many species in phy object? 94
length(species_phy$tip.label)

## Import habitat metadata from FishBase
species_list <- tolower(gsub("_", " ", species_phy$tip.label))
  
species_dist <- data.frame(tip.label = tolower(distribution(species_list(Family = 
                                                                           c("Zoarcidae", "Anarhichadidae",
                                                                             "Bathymasteridae", "Pholidae",
                                                                             "Stichaeidae", "Gasterosteidae")))$Species),
                           Region = distribution(species_list(Family = c("Zoarcidae", "Anarhichadidae",
                                                                             "Bathymasteridae", "Pholidae",
                                                                             "Stichaeidae", "Gasterosteidae")))$FAO,
                           Lat = distribution(species_list(Family = c("Zoarcidae", "Anarhichadidae",
                                                                             "Bathymasteridae", "Pholidae",
                                                                             "Stichaeidae", "Gasterosteidae")))$LatDeg,
                           NS = distribution(species_list(Family = c("Zoarcidae", "Anarhichadidae",
                                                                             "Bathymasteridae", "Pholidae",
                                                                             "Stichaeidae", "Gasterosteidae")))$N_S)
distribution(species_list(Family = "Zoarcidae"))

# Create variable for whether row is Antarctic, Arctic, polar
species_dist$Polar <- ifelse(grepl("Arctic", species_dist$Region) == TRUE, "Arctic",
                             ifelse(grepl("Antarctic", species_dist$Region) == TRUE, "Antarctic",
                                    "Subpolar"))

## Combine latitude with Zoarcid phylogeny object and plot
species_phy$tip.label <- tolower(gsub("_", " ", species_phy$tip.label))

species_dist_filt <- filter(species_dist, tip.label %in% species_phy$tip.label)

#Summarize latitude
species_dist_sum <- summarySE(measurevar = "Lat",
                              groupvars = c("tip.label", "Polar"),
                              data = species_dist_filt)

# Colored plot
max_y <- max(species_phy$edge.length)

ggtree(species_phy, layout = "rectangular", 
       aes(color = Lat), size = .75) %<+% species_dist_sum +
  geom_tippoint(aes(shape = Polar), size = 3) +
  geom_tiplab(hjust = -0.05) +
  scale_shape_manual(values = c(19,1,NA), na.translate = F) +
  theme_tree() +
  geom_treescale(width = 20) +
  scale_color_viridis_c(direction = -1) +
  theme_classic(base_size = 20) +
  xlim(0, 100) +
  theme(legend.background = element_rect(fill = "transparent"),
        panel.background = element_rect(fill = "transparent"),
        strip.background = element_rect(fill = "transparent"),
        plot.background = element_rect(fill = "transparent", color = NA),
        axis.line = element_blank(),
        axis.text = element_blank(),
        axis.ticks = element_blank(),
        legend.text = element_text(size = 8)) +
  labs(color = "Latitude", shape = "Region") +
  geom_rect(xmin = max_y-2, xmax = max_y-3, ymin = -Inf, ymax = Inf, fill = "skyblue", alpha = 0.01, lty = 0) +
  geom_rect(xmin = max_y-10, xmax = max_y-15, ymin = -Inf, ymax = Inf, fill = "pink", alpha = 0.01, lty = 0)

```

