---
title: "AFP_LongRead_Notebook"
author: "Sam Bogan and Nathan Surendran"
output:
  github_document: default
---

# Summary

This is an R markdown lab notebook for all research, wet lab work, and analyses related to a study on type AFP III evolution revealed by long read sequencing and genome assembly in Zoarcoid fishes. The 11 Zoarcoid species included in the study and Scorpaeiformes outgroups (sticklebacks) are indexed

File registry
1. Metadata
  + LongRead_SpeciesIDs.csv (index of species names and genome IDs)

\br

Below is a quick visual representation of the species' relatedness using the rfishtree time-calibrated phylogeny

```{r setup, include=FALSE}

knitr::opts_knit$set(root.dir = '~/Documents/GitHub/Long_AFP/')

# Load packages for plotting quick species tree
library(tidyverse)
library(fishtree)
library(ggtree)
library(MatrixModels)
library(rfishbase)
library(phytools)
library(ape)

```

What genomes are we working with?

```{r}

# Read in list of species
Species_df <- read.csv("LongRead_SpeciesIDs.csv")

# Print list to show which genomes came from public source vs. Kelley Lab
head(Species_df)

```

Extract and plot species phylogeny with important metadata

```{r}

# Extract eelpout phylogeny
species_phy <- fishtree_phylogeny(species = Species_df$Species_ID, type=c("phylogram"))

# How many species in phy object? 94
length(species_phy$tip.label)

# Plot multilocus phylogeny
ggtree(species_phy, layout = "rectangular", size = .75) +
  geom_tiplab(hjust = -0.05) +
  scale_shape_manual(values = c(19,1,NA), na.translate = F) +
  xlim(0, .75) +
  theme_tree() +
  labs(title = "Multilocus Rabosky FishTreeofLife phylogeny")

```

Make phylogeny w/ metadata

```{r, eval = FALSE}
## Import habitat metadata from FishBase
species_list <- tolower(gsub("_", " ", species_phy$tip.label))
  
species_dist <- data.frame(tip.label = tolower(distribution(species_list(Family = 
                                                                           c("Zoarcidae", "Anarhichadidae",
                                                                             "Bathymasteridae", "Pholidae",
                                                                             "Stichaeidae", "Gasterosteidae")))$Species),
                           Region = distribution(species_list(Family = c("Zoarcidae", "Anarhichadidae",
                                                                             "Bathymasteridae", "Pholidae",
                                                                             "Stichaeidae", "Gasterosteidae")))$FAO,
                           Lat = distribution(species_list(Family = c("Zoarcidae", "Anarhichadidae",
                                                                             "Bathymasteridae", "Pholidae",
                                                                             "Stichaeidae", "Gasterosteidae")))$LatDeg,
                           NS = distribution(species_list(Family = c("Zoarcidae", "Anarhichadidae",
                                                                             "Bathymasteridae", "Pholidae",
                                                                             "Stichaeidae", "Gasterosteidae")))$N_S)
distribution(species_list(Family = "Zoarcidae"))

# Create variable for whether row is Antarctic, Arctic, polar
species_dist$Polar <- ifelse(grepl("Arctic", species_dist$Region) == TRUE, "Arctic",
                             ifelse(grepl("Antarctic", species_dist$Region) == TRUE, "Antarctic",
                                    "Subpolar"))

## Combine latitude with Zoarcid phylogeny object and plot
species_phy$tip.label <- tolower(gsub("_", " ", species_phy$tip.label))

species_dist_filt <- filter(species_dist, tip.label %in% species_phy$tip.label)

#Summarize latitude
species_dist_sum <- summarySE(measurevar = "Lat",
                              groupvars = c("tip.label", "Polar"),
                              data = species_dist_filt)

# Colored plot
max_y <- max(species_phy$edge.length)

ggtree(species_phy, layout = "rectangular", 
       aes(color = Lat), size = .75) %<+% species_dist_sum +
  geom_tippoint(aes(shape = Polar), size = 3) +
  geom_tiplab(hjust = -0.05) +
  scale_shape_manual(values = c(19,1,NA), na.translate = F) +
  theme_tree() +
  geom_treescale(width = 20) +
  scale_color_viridis_c(direction = -1) +
  theme_classic(base_size = 20) +
  xlim(0, 100) +
  theme(legend.background = element_rect(fill = "transparent"),
        panel.background = element_rect(fill = "transparent"),
        strip.background = element_rect(fill = "transparent"),
        plot.background = element_rect(fill = "transparent", color = NA),
        axis.line = element_blank(),
        axis.text = element_blank(),
        axis.ticks = element_blank(),
        legend.text = element_text(size = 8)) +
  labs(color = "Latitude", shape = "Region") +
  geom_rect(xmin = max_y-2, xmax = max_y-3, ymin = -Inf, ymax = Inf, fill = "skyblue", alpha = 0.01, lty = 0) +
  geom_rect(xmin = max_y-10, xmax = max_y-15, ymin = -Inf, ymax = Inf, fill = "pink", alpha = 0.01, lty = 0)

```

# Species tree construction

### 10/26/2023 Running BUCSO on long read assemblies

NS: Ran a BUSCO analysis on the fasta genome data for eelpouts and zoarcoids.  Single seq files then compressed using tar.gz

The slurm job script below was ran for the following species:
* alupus
* aminor
* bsig
* byu
* cviol1
* gacul
* lmac1
* lpac
* melgel1
* norway
* oamb
* pgunn1

```{bash, eval = FALSE}


#!/bin/bash
#SBATCH --job-name=busco_assembly_stats_alupus
#SBATCH --time=0-12:00:00
#SBATCH --partition=Instruction
#SBATCH --mail-user=nsurendr@ucsc.edu
#SBATCH --mail-type=ALL
#SBATCH --output=busco_assembly.out
#SBATCH --error=busco_assembly.err
#SBATCH --ntasks=1
#SBATCH --ntasks-per-node=1
#SBATCH --cpus-per-task=16
#SBATCH --mem=6GB

module load hb hb-gnu busco/busco-5.4.7
busco -c 16 -i /hb/groups/kelley_training/assemblies/alupus.asm.p_ctg.fa -o busco_assembly_alupus --auto-lineage -m genome -f

```

### 10/29/2023 Running BUSCO_phylogenomics to create multiple gene trees

SB ran reran BUSCO on the 12 assemblies above using his HB allocation and input the BUSCO output to BUSCO_to_phylogeny pipeline described here: https://github.com/jamiemcg/BUSCO_phylogenomics

NS's BUSCO job's per species were moved to Archive directory

The job script below has an annotation for where NS would start the script since he has already run BUSCO on the 12 assemblies

Before running the job, install BUSCO_phylogenomics and its dependencies

```{bash, eval = FALSE}

# Clone and enter repo
git clone https://github.com/jamiemcg/BUSCO_phylogenomics

# Create supplementary conda environment
module load miniconda3.9

conda env create -p BUSCO_phylogenomics_supp

conda activate BUSCO_phylogenomics_supp

# Install the following packages
mamba install biopython
conda install -c bioconda muscle
conda install -c bioconda trimal
conda install -c bioconda fasttree
conda install -c bioconda iqtree

```

Here is the job script

```{bash, eval = FALSE}

#!/bin/bash
#SBATCH --job-name=BUSCO_to_phy
#SBATCH --time=0-48:00:00
#SBATCH --mail-user=snbogan@ucsc.edu
#SBATCH --mail-type=ALL
#SBATCH --output=BUSCO_to_phy.out
#SBATCH --error=BUSCO_to_phy.err
#SBATCH --ntasks=1
#SBATCH --ntasks-per-node=1
#SBATCH --cpus-per-task=16
#SBATCH --mem=12GB

# Load BUSCO and parallel
module load busco
module load parallel

# Go to home directory
cd /hb/home/snbogan/

# Make BUSCO_results directory
mkdir /hb/home/snbogan/PolarFish/Long_AFP/BUSCO_results

# Go to folder with .fa genomes
cd /hb/home/snbogan/PolarFish/Long_AFP/

# Run BUSCO in parallel
run_busco() {
    name=$(basename "$1" .fa)
    busco -c 8 -i "$name".fa -o busco_assembly_"$name" --auto-lineage -m genome -f
}

export -f run_busco

find . -maxdepth 1 -name "*.fa" | parallel -j2 run_busco {}

##################################
###### NS should start here ######
##################################

# Activate BUSCO_phylogenomics conda env
module load miniconda3.9
conda activate /hb/home/snbogan/BUSCO_phylogenomics_supp

# Go to working directory
cd /hb/home/snbogan/PolarFish/Long_AFP/

# Run pipeline command
python /hb/home/snbogan/BUSCO_phylogenomics/BUSCO_phylogenomics.py \
 -i BUSCO_results -o output_busco_phylogenomics -t 8

## If data look patchy, run python <python count_buscos.py -i BUSCO_runs>

```

### 10/31/2023 Creating species tree from supermatrix alignment

SB used IQtree to create a species tree from the BUSCO_phylogenomics supermatrix alignment. This was a first pass at creating a species tree. NS will reproduce the BUSCO_phylogenomics and IQtree analyses and then move forward on creating a consensus species tree from the BUSCO_phylogenomics output of multiple gene trees. This will also use IQtree. SB's supermatrix IQtree code is below.

```{bash, eval = FALSE}

#!/bin/bash
#SBATCH --job-name=IQtree_supermatrix
#SBATCH --time=0-12:00:00
#SBATCH --mail-user=snbogan@ucsc.edu
#SBATCH --mail-type=ALL
#SBATCH --output=IQtree_supermatrix.out
#SBATCH --error=IQtree_supermatrix.err
#SBATCH --ntasks=1
#SBATCH --ntasks-per-node=1
#SBATCH --cpus-per-task=8
#SBATCH --mem=12GB

# Activate BUSCO_phylogenomics conda env
module load miniconda3.9
conda activate /hb/home/snbogan/BUSCO_phylogenomics_supp

# Go to BUSCO_phylogenomics directory w/ supermatrix alignment
cd /hb/home/snbogan/PolarFish/Long_AFP/output_busco_phylogenomics/supermatrix

# Run phyml on supermatrix alignment
iqtree -s SUPERMATRIX.phylip -o busco_assembly_Gacul_UGA_v5_genomic -m WAG -redo

```

IQtree finished but the distance matrix and tree files do not seem to align. Something's up.

```{r}

# Read and print IQtree distance matrix
mldist <- as.data.frame(
  read.table("Trees/Supermatrix_IQtree_LG4M/SUPERMATRIX.phylip.txt")
  )

row.names(mldist) <- mldist$V1

mldist_matrix <- as.matrix(mldist[,-c(1)])

hc <- hclust(as.dist(mldist_matrix))

# Convert hierarchical clustering tree to a phylogenetic tree
dist_tree <- as.phylo(hc)

# Plot distance tree
ggtree(dist_tree, layout = "rectangular") +
  geom_tiplab() +
  geom_treescale(width = 0.10, x = 0) +
  xlim(0, .15) +
  labs(title = "IQtree distance tree")

# Read tree file
tree <- read.tree("Trees/Supermatrix_IQtree_LG4M/SUPERMATRIX.phylip.treefile")

tree$tip.label <- gsub("busco_assembly_", "", tree$tip.label)

tree <- reroot(tree,
               node.number = 6,
               position = .5*(max(tree$edge.length)))

ggtree(tree, layout = "rectangular") +
  geom_tiplab() +
  xlim(0, .125) +
  geom_treescale() +
  labs(title = "Rooted IQtree LG4M substitution model")

```

### 11/1/2023

SB reran IQtree job replacing substitution model w/ WAG. The goal is to simplify the model and hopefully resolve a better variable rate tree. It produced another faulty tree.

Next, SB rooted the tree by adding the parameter. The result looked better.

```{r}

# Read tree file
tree <- read.newick("Trees/Supermatrix_IQtree_WAGroot/SUPERMATRIX.phylip.treefile")

tree$tip.label <- gsub("busco_assembly_", "", tree$tip.label)

tree <- reroot(tree, 
             node.number = 12, 
             position = .5*(max(tree$edge.length)))

tree$tip.label <- c("Anarhichas minor", "Anarhichas lupus", "Lycodes diapterus", 
                    "Ophthalmolycus amberensis", "Lycodes pacificus", "Melanostigma gelatinosum",
                    "Lycodes platyrhinus", "Leptoclinus maculatus", "Pholis gunnellus", 
                    "Cebidichthys violaceus", "Bathymaster signatus", "Gasterosteus aculeatus")

busco_df <- read.csv(("Busco_results_genome_size.csv"))

# Plot
Fig_1A <- ggtree(tree, layout = "rectangular", size = 1.5) %<+% busco_df +
  geom_tiplab(color = "black", nudge_x = .005) +
  geom_tippoint(aes(size = log(Genome_size))) +
  geom_treescale(width = 0.025, offset = -.55) +
  scale_color_viridis_c(direction = -1) +
  theme(legend.position = "none") +
  xlim(0, .15)

# Export Fig 1B as png
png("~/Documents/GitHub/Long_AFP/Figures/Fig_1A.png", units = "in", width = 8, 
    height = 6, 
    res = 600)

Fig_1A
       


```


### 11/7/2023

Meeting w/ Ella Gustavson on reverse amplicon identification in Norway eelpout using PCR primers

Goal: We are going to automate Ella's amplicon ID'ing process during a 11/9/2023 meeting using an R script

Below is a list of Ella's steps for amplicon ID starting with forward and reverse primer set

Alignment
1. Create blast databases w/ parsed seqIDs
2. Two blasts: one with forward, one w/ reverse primer
2. From blast output, create table of hits and alignment metrics

### 11/15/2023

Creating multigene coalescence tree using Astral

```{bash, eval = FALSE}

#!/bin/bash
#SBATCH --job-name=astral_tree
#SBATCH --mail-user=snbogan@ucsc.edu
#SBATCH --mail-type=ALL
#SBATCH --output=astral_tree.out
#SBATCH --error=astral_tree.err
#SBATCH --time=01:00:00
#SBATCH --ntasks=1
#SBATCH --ntasks-per-node=1
#SBATCH --cpus-per-task=8
#SBATCH --mem=12GB

# Move to working directory
cd /hb/home/snbogan/PolarFish/Long_AFP/output_busco_phylogenomics/gene_trees_single_copy

# Load java
module load java

# Run ASTRAL to create the multigene consensus tree
java -jar /hb/home/snbogan/Astral/astral.5.7.8.jar \
-i /hb/home/snbogan/PolarFish/Long_AFP/output_busco_phylogenomics/gene_trees_single_copy/ALL.tree \
-o /hb/home/snbogan/PolarFish/Long_AFP/output_busco_phylogenomics/gene_trees_single_copy/ASTRAL_out.tree \
2>ASTRAL_out.log

echo "ASTRAL analysis complete."

```

Plot unrooted astral coalescence tree

```{r}

# Read in astral output tree
astral_tree <- read.tree("Trees/Astral/ASTRAL_out.tree")

# ggplot
astral_tree$tip.label <- gsub("busco_assembly_", "", astral_tree$tip.label)

astral_tree$edge.length <- ifelse(is.na(astral_tree$edge.length) == TRUE, 
                                  1, astral_tree$edge.length)

astral_tree <- reroot(astral_tree,
                      node.number = 1,
                      position = .5*(max(tree$edge.length)))

ggtree(astral_tree, layout = "rectangular") +
  geom_tiplab() + 
  geom_nodelab(vjust = -.75, hjust = 1.5, color = "blue") +
  geom_treescale(width = 1, offset = -.55) +
  xlim(0, 12) +
  labs(title = "Rooted astral multigene coalescence tree")

```
Looks like IQtree performed well. Let's move forward with creating a time-calibrated tree.

### 11/30/2023

A .Rmd for figures 1 and 2 of the manuscript was created under Figures.Rmd

# Gene tree construction

### 12/13/2023

Hand annotation of AFP regions was performed by SB using two fasta files of zoarcoid AFPs using the following blast parameters mimicking NCBI blastn 'more dissimilar sequences/discontinguous blast'.

AFPs were documented as putatively complete genes if they contained the following:

All blast outputs can be found in ~/Documents/GitHub/Long_AFP/Hand_Annot/Blasts

* Start codon
* Complete exons 1 and 2

```{bash, eval = FALSE}

blastn -query /path/to/query.fa \
       -subject /path/to/genome.fa \
       -out output.txt \
       -evalue 0.1 \
       -word_size 11 \
       -gapopen 5 \
       -gapextend 2 \
       -penalty -3 \
       -reward 2 \
       -dust yes \
       -soft_masking true \
       -task blastn

```

The resulting annotations are in the file AFP_SASA_SASB_ranges.csv and can be seen below:

```{r}

# Read in hand annotation csv and print
head(read.csv("AFP_SASA_SASB_ranges.csv"), n = 50)

```

The AFP fasta's used for AFP annotations were:

* J03924.1: Macrozoarces americanus antifreeze protein (OP5A) gene, complete cds (primary)
* JQ040521.1: Anarhichas lupus clone AWG1 type III antifreeze protein (AFP III) gene, complete cds

JQ040521.1 was used to double check species for which no apparent AFPs were annotated

### 01/02/2024

Hand annotations of SAS-A and SAS-B were performed by SB using G. aculeatus paralogs:

* NC_053220.1:c8556783-8553001 (SAS-A)
* NC_053220.1:c8562485-8557311 (SAS-B)

SAS-B was identified based on its homology to Zoarcoid AFP III orthologs

All blast outputs can be found in ~/Documents/GitHub/Long_AFP/Hand_Annot/Blasts

### 01/04/2024

A gene tree was constructed by converting hand annotations of SAS and AFP regions to .bed files by hand and extracting their underlying sequences using bedtools getfasta. FASTA files were then concatenated into one file and aligned using MAFFT and constructed into a phylogenetic tree using IQTREE.

Below is the code for FASTA extraction and concatenation

```{bash, eval = FALSE}

# Extract fasta sequencing using .bed coordinates
bedtools getfasta -fi /path/to/genome.fa \ 
-bed /path/to/coordinates.bed \
-fo /path/to/output.fa # Repeat across all SAS/AFP coordinates

```

```{bash, eval = FALSE}
#!/bin/bash

## Combine all fasta's using concat.sh in 'Scripts' folder
# Target directory containing .fa files
TARGET_DIR="/Users/sambogan/Documents/GitHub/Long_AFP/Hand_Annot/GetFasta"

# Output file
OUTPUT_FILE="combined.fa"

# Navigate to target directory
cd "$TARGET_DIR"

# Check if output file already exists
if [ -f "$OUTPUT_FILE" ]; then
    echo "Output file $OUTPUT_FILE already exists. Removing it."
    rm "$OUTPUT_FILE"
fi

# Concatenate all .fa files into one
for file in *.fa; do
    cat "$file" >> "$OUTPUT_FILE"
    echo "Concatenated $file"
done

echo "All .fa files have been concatenated into $OUTPUT_FILE."

```

The resulting concatentation of fasta sequences was aligned using MAFFT set to preset parameters at: https://www.ebi.ac.uk/Tools/msa/mafft/

The .fa alignment file was input to IQTREE set to automated substitution model selection and free rate variation = FALSE. The selected substitution model was K3Pu+F+G4.

The resulting IQTREE gene tree is stored under Trees/Gene_Trees/01042024_IQTREE_K3Pu+F+G4_SAS_AFP_genetree.newick and is shown here:

```{r}

# Read in IQTree gene tree
IQ_genetree <- read.tree("Trees/Gene_Trees/01042024_IQTREE_K3Pu+F+G4_SAS_AFP_genetree.newick")

# Create factor for AFPs and SAS genes
group_df <- data.frame(tip.label = IQ_genetree$tip.label,
  group = ifelse(
  grepl("SAS", IQ_genetree$tip.label), "SAS",
  ifelse(grepl("AFP", IQ_genetree$tip.label), "AFP", NA)))

IQ_genetree <- reroot(IQ_genetree, 
             node.number = 193, 
             position = .5*(max(tree$edge.length)))

# Plot tree
ggtree(IQ_genetree, aes(color = group), layout = "rectangular") %<+% group_df +
  geom_tiplab(size = 2, align = TRUE) +
  geom_treescale() +
  labs(title = "Rooted IQtree gene tree K3Pu+F+G4 substitution model")


```

As shown above, this preliminary analysis demonstrates two indpendent origins of AFP III genes arising from SAS. This is highly unlikely given the prior literature. The most derived clade of SAS genes should be inspected. The following explanations may explain this result:

* Derived SAS genes were incorrectly annotated as AFPs and only possess first and last exon of SAS-B.
* These derived SAS genes underwent relaxed constraint resemble AFPs in terms of their conserved exons 

Additionally, some sorting of SAS-A and SAS-B appears incorrect and needs to be inspected as well.

