---
title: "AFP_LongRead_Notebook"
author: "Sam Bogan and Nathan Surendran"
output:
  github_document: default
---

#Summary

This is an R markdown lab notebook for all research, wet lab work, and analyses related to a study on type AFP III evolution revealed by long read sequencing and genome assembly in Zoarcoid fishes. The 11 Zoarcoid species included in the study and Scorpaeiformes outgroups (sticklebacks) are indexed

File registry
1. Metadata
  + LongRead_SpeciesIDs.csv (index of species names and genome IDs)

\br

Below is a quick visual representation of the species' relatedness using the rfishtree time-calibrated phylogeny

```{r setup, include=FALSE}

knitr::opts_knit$set(root.dir = '~/Documents/GitHub/Long_AFP/')

# Load packages for plotting quick species tree
library(tidyverse)
library(fishtree)
library(ggtree)
library(MatrixModels)
library(rfishbase)
library(phytools)
library(ape)

```

What genomes are we working with?

```{r}

# Read in list of species
Species_df <- read.csv("LongRead_SpeciesIDs.csv")

# Print list to show which genomes came from public source vs. Kelley Lab
head(Species_df)

```

Extract and plot species phylogeny with important metadata

```{r}

# Extract eelpout phylogeny
species_phy <- fishtree_phylogeny(species = Species_df$Species_ID, type=c("chronogram"))

# How many species in phy object? 94
length(species_phy$tip.label)

# Plot multilocus phylogeny
ggtree(species_phy, layout = "rectangular", size = .75) +
  geom_tiplab(hjust = -0.05) +
  scale_shape_manual(values = c(19,1,NA), na.translate = F) +
  xlim(0, 90) +
  theme_tree() +
  labs(title = "Multilocus Rabosky FishTreeofLife phylogeny")

```

Make phylogeny w/ metadata

```{r, eval = FALSE}
## Import habitat metadata from FishBase
species_list <- tolower(gsub("_", " ", species_phy$tip.label))
  
species_dist <- data.frame(tip.label = tolower(distribution(species_list(Family = 
                                                                           c("Zoarcidae", "Anarhichadidae",
                                                                             "Bathymasteridae", "Pholidae",
                                                                             "Stichaeidae", "Gasterosteidae")))$Species),
                           Region = distribution(species_list(Family = c("Zoarcidae", "Anarhichadidae",
                                                                             "Bathymasteridae", "Pholidae",
                                                                             "Stichaeidae", "Gasterosteidae")))$FAO,
                           Lat = distribution(species_list(Family = c("Zoarcidae", "Anarhichadidae",
                                                                             "Bathymasteridae", "Pholidae",
                                                                             "Stichaeidae", "Gasterosteidae")))$LatDeg,
                           NS = distribution(species_list(Family = c("Zoarcidae", "Anarhichadidae",
                                                                             "Bathymasteridae", "Pholidae",
                                                                             "Stichaeidae", "Gasterosteidae")))$N_S)
distribution(species_list(Family = "Zoarcidae"))

# Create variable for whether row is Antarctic, Arctic, polar
species_dist$Polar <- ifelse(grepl("Arctic", species_dist$Region) == TRUE, "Arctic",
                             ifelse(grepl("Antarctic", species_dist$Region) == TRUE, "Antarctic",
                                    "Subpolar"))

## Combine latitude with Zoarcid phylogeny object and plot
species_phy$tip.label <- tolower(gsub("_", " ", species_phy$tip.label))

species_dist_filt <- filter(species_dist, tip.label %in% species_phy$tip.label)

#Summarize latitude
species_dist_sum <- summarySE(measurevar = "Lat",
                              groupvars = c("tip.label", "Polar"),
                              data = species_dist_filt)

# Colored plot
max_y <- max(species_phy$edge.length)

ggtree(species_phy, layout = "rectangular", 
       aes(color = Lat), size = .75) %<+% species_dist_sum +
  geom_tippoint(aes(shape = Polar), size = 3) +
  geom_tiplab(hjust = -0.05) +
  scale_shape_manual(values = c(19,1,NA), na.translate = F) +
  theme_tree() +
  geom_treescale(width = 20) +
  scale_color_viridis_c(direction = -1) +
  theme_classic(base_size = 20) +
  xlim(0, 100) +
  theme(legend.background = element_rect(fill = "transparent"),
        panel.background = element_rect(fill = "transparent"),
        strip.background = element_rect(fill = "transparent"),
        plot.background = element_rect(fill = "transparent", color = NA),
        axis.line = element_blank(),
        axis.text = element_blank(),
        axis.ticks = element_blank(),
        legend.text = element_text(size = 8)) +
  labs(color = "Latitude", shape = "Region") +
  geom_rect(xmin = max_y-2, xmax = max_y-3, ymin = -Inf, ymax = Inf, fill = "skyblue", alpha = 0.01, lty = 0) +
  geom_rect(xmin = max_y-10, xmax = max_y-15, ymin = -Inf, ymax = Inf, fill = "pink", alpha = 0.01, lty = 0)

```

# HB SLURM Scripts

### 10/26/2023 Running BUCSO on long read assemblies

NS: Ran a BUSCO analysis on the fasta genome data for eelpouts and zoarcoids.  Single seq files then compressed using tar.gz

The slurm job script below was ran for the following species:
* alupus
* aminor
* bsig
* byu
* cviol1
* gacul
* lmac1
* lpac
* melgel1
* norway
* oamb
* pgunn1

```{bash, eval = FALSE}


#!/bin/bash
#SBATCH --job-name=busco_assembly_stats_alupus
#SBATCH --time=0-12:00:00
#SBATCH --partition=Instruction
#SBATCH --mail-user=nsurendr@ucsc.edu
#SBATCH --mail-type=ALL
#SBATCH --output=busco_assembly.out
#SBATCH --error=busco_assembly.err
#SBATCH --ntasks=1
#SBATCH --ntasks-per-node=1
#SBATCH --cpus-per-task=16
#SBATCH --mem=6GB

module load hb hb-gnu busco/busco-5.4.7
busco -c 16 -i /hb/groups/kelley_training/assemblies/alupus.asm.p_ctg.fa -o busco_assembly_alupus --auto-lineage -m genome -f

```

### 10/29/2023 Running BUSCO_phylogenomics to create multiple gene trees

SB ran reran BUSCO on the 12 assemblies above using his HB allocation and input the BUSCO output to BUSCO_to_phylogeny pipeline described here: https://github.com/jamiemcg/BUSCO_phylogenomics

NS's BUSCO job's per species were moved to Archive directory

The job script below has an annotation for where NS would start the script since he has already run BUSCO on the 12 assemblies

Before running the job, install BUSCO_phylogenomics and its dependencies

```{bash, eval = FALSE}

# Clone and enter repo
git clone https://github.com/jamiemcg/BUSCO_phylogenomics

# Create supplementary conda environment
module load miniconda3.9

conda env create -p BUSCO_phylogenomics_supp

conda activate BUSCO_phylogenomics_supp

# Install the following packages
mamba install biopython
conda install -c bioconda muscle
conda install -c bioconda trimal
conda install -c bioconda fasttree
conda install -c bioconda iqtree

```

Here is the job script

```{bash, eval = FALSE}

#!/bin/bash
#SBATCH --job-name=BUSCO_to_phy
#SBATCH --time=0-48:00:00
#SBATCH --mail-user=snbogan@ucsc.edu
#SBATCH --mail-type=ALL
#SBATCH --output=BUSCO_to_phy.out
#SBATCH --error=BUSCO_to_phy.err
#SBATCH --ntasks=1
#SBATCH --ntasks-per-node=1
#SBATCH --cpus-per-task=16
#SBATCH --mem=12GB

# Load BUSCO and parallel
module load busco
module load parallel

# Go to home directory
cd /hb/home/snbogan/

# Make BUSCO_results directory
mkdir /hb/home/snbogan/PolarFish/Long_AFP/BUSCO_results

# Go to folder with .fa genomes
cd /hb/home/snbogan/PolarFish/Long_AFP/

# Run BUSCO in parallel
run_busco() {
    name=$(basename "$1" .fa)
    busco -c 8 -i "$name".fa -o busco_assembly_"$name" --auto-lineage -m genome -f
}

export -f run_busco

find . -maxdepth 1 -name "*.fa" | parallel -j2 run_busco {}

##################################
###### NS should start here ######
##################################

# Activate BUSCO_phylogenomics conda env
module load miniconda3.9
conda activate /hb/home/snbogan/BUSCO_phylogenomics_supp

# Go to working directory
cd /hb/home/snbogan/PolarFish/Long_AFP/

# Run pipeline command
python /hb/home/snbogan/BUSCO_phylogenomics/BUSCO_phylogenomics.py \
 -i BUSCO_results -o output_busco_phylogenomics -t 8

## If data look patchy, run python <python count_buscos.py -i BUSCO_runs>

```

### 10/31/2023 Creating species tree from supermatrix alignment

SB used IQtree to create a species tree from the BUSCO_phylogenomics supermatrix alignment. This was a first pass at creating a species tree. NS will reproduce the BUSCO_phylogenomics and IQtree analyses and then move forward on creating a consensus species tree from the BUSCO_phylogenomics output of multiple gene trees. This will also use IQtree. SB's supermatrix IQtree code is below.

```{bash, eval = FALSE}

#!/bin/bash
#SBATCH --job-name=IQtree_supermatrix
#SBATCH --time=0-12:00:00
#SBATCH --mail-user=snbogan@ucsc.edu
#SBATCH --mail-type=ALL
#SBATCH --output=IQtree_supermatrix.out
#SBATCH --error=IQtree_supermatrix.err
#SBATCH --ntasks=1
#SBATCH --ntasks-per-node=1
#SBATCH --cpus-per-task=8
#SBATCH --mem=12GB

# Activate BUSCO_phylogenomics conda env
module load miniconda3.9
conda activate /hb/home/snbogan/BUSCO_phylogenomics_supp

# Go to BUSCO_phylogenomics directory w/ supermatrix alignment
cd /hb/home/snbogan/PolarFish/Long_AFP/output_busco_phylogenomics/supermatrix

# Run phyml on supermatrix alignment
iqtree -s SUPERMATRIX.phylip -m WAG # SB used LG4M

```

IQtree finished but the distance matrix and tree files do not seem to align. Something's up.

```{r}

# Read and print IQtree distance matrix
mldist <- as.data.frame(
  read.table("Trees/Supermatrix_IQtree_LG4M/SUPERMATRIX.phylip.txt")
  )

row.names(mldist) <- mldist$V1

mldist_matrix <- as.matrix(mldist[,-c(1)])

hc <- hclust(as.dist(mldist_matrix))

# Convert hierarchical clustering tree to a phylogenetic tree
dist_tree <- as.phylo(hc)

# Plot distance tree
ggtree(dist_tree, layout = "rectangular") +
  geom_tiplab() +
  geom_treescale(width = 0.10, x = 0) +
  xlim(0, .15) +
  labs(title = "IQtree distance tree")

# Read tree file
tree <- read.tree("Trees/Supermatrix_IQtree_LG4M/SUPERMATRIX.phylip.treefile")

ggtree(tree, layout = "rectangular") +
  geom_tiplab() +
  xlim(0, .25) +
  geom_treescale() +
  labs(title = "Terrible looking IQtree LG4M substitution model")

```

### 11/1/2023

SB reran IQtree job replacing substitution model w/ WAG. The goal is to simplify the model and hopefully resolve a better variable rate tree. It produced another faulty tree.

Next, SB rooted the tree by adding the parameter. The result looked better.

```{r}

# Read tree file
tree <- read.tree("Trees/Supermatrix_IQtree_WAGroot/SUPERMATRIX.phylip.treefile")

ggtree(tree, layout = "rectangular") +
  geom_tiplab() +
  geom_treescale() +
  xlim(0, .15) +
  labs(title = "Rooted IQtree WAG substitution model")

```


### 11/7/2023

Meeting w/ Ella Gustavson on reverse amplicon identification in Norway eelpout using PCR primers

Goal: We are going to automate Ella's amplicon ID'ing process during a 11/9/2023 meeting using an R script

Below is a list of Ella's steps for amplicon ID starting with forward and reverse primer set

Alignment
1. Create blast databases w/ parsed seqIDs
2. Two blasts: one with forward, one w/ reverse primer
2. From blast output, create table of hits and alignment metrics

### 11/9/2023

