---
title: "Figures"
author: "Sam Bogan"
date: "2023-11-30"
output: html_document
---

```{r setup, include=FALSE}

knitr::opts_knit$set(root.dir = '~/Documents/GitHub/Long_AFP/')

# Load packages for plotting quick species tree
library(tidyverse)
library(fishtree)
library(ggtree)
library(MatrixModels)
library(rfishbase)
library(phytools)
library(ape)
library(viridis)
library(ggpubr)
library(Rmisc)
library(rnaturalearth)
library(rnaturalearthdata)
library(reshape2)
library(sf)
library(geodata)
library(data.table)
library(ggrepel)

```

##Figure 1 - Ranges, depths, species tree, genome QC metrics

###Panel A - range, depth, temperature

```{r}

# Read in SST data
sst <- read.csv(("ncdcOisst2Agg_6cc2_4851_60b3.csv"))
sst$sst <- as.numeric(gsub("NaN", NA, sst$sst))
sst <- na.omit(sst)

# Calc mean SST per coordinate
sst_sum <- summarySE(measurevar = "sst",
                     groupvars = c("latitude", "longitude"),
                     data = sst)

# Fix logitude object
sst_sum$longitude <- sst_sum$longitude - 360

sst_sum$longitude <- ifelse(sst_sum$longitude < -180,
                            sst_sum$longitude + 360,
                            sst_sum$longitude)

# Create data for plotting
df <- sst_sum %>%
  st_as_sf(coords = c("longitude", "latitude"), crs = 4326)

world <- ne_countries(scale = "medium", returnclass = "sf")

# Partial plot
Fig1A <- ggplot() +
  geom_sf(data = world, fill = "grey20", color = "black") +
  geom_sf(data = df, aes(color = sst), alpha = 0.7, size = 0.1) +
  theme_classic(base_size = 20) +
  theme(legend.position = "top") +
  scale_color_viridis_c() +
  labs(color = "SST (Â°C)") +
  theme(axis.title = element_blank())

Fig1A

Species_List <- read.csv("LongRead_SpeciesIDs.csv")

```

Plot most likely occurences derived from AquaMaps range data

```{r}

# Move to input data directory for tempm time series
setwd("Range_Data/AquaMaps/")

# Get the files names
AQUAs <- list.files(pattern = "*.csv")

# Read in .csv files listed in temps
AQUA_dfs <- lapply(AQUAs, read.csv)

all_AQUA_df <- rbindlist(AQUA_dfs,
                         idcol = "species")

all_AQUA_df$species <- paste(all_AQUA_df$Genus,
                             all_AQUA_df$Species,
                             sep = " ")

colnames(all_AQUA_df) <- c('species','Genus','Species',
                           'latitude', 'longitude', 'C.Square.Code', 
                           'Overall.Probability')

# Color plot by max depth
depth_df <- fb_tbl("species")

depth_df$species <- paste(depth_df$Genus,
                          depth_df$Species,
                          sep = " ")

depth_df <- filter(depth_df, species %in% Species_List$Species_ID)

depth_sum <- summarySE(measurevar = "DepthRangeDeep",
                       groupvars = c("species"),
                       data = depth_df)

centroids <- all_AQUA_df %>%
            dplyr::group_by(Species) %>%
            dplyr::slice(which.max(Overall.Probability))

centroids <- merge(centroids, depth_sum, by = "species")

sum_range <- all_AQUA_df %>%
  dplyr::group_by(species) %>%
  dplyr::summarise(sum = sum(Overall.Probability))

centroids <- merge(centroids, sum_range, by = "species")

centroids <- rbind(centroids,
                   data.frame(species = "Lycodes platyrhinus",
                              Genus = "Lycodes",
                              Species = "platyrhinus",
                              latitude = 73.8,
                              longitude = 6.9,
                              C.Square.Code = NA,
                              Overall.Probability = NA,
                              N = 4,
                              DepthRangeDeep = NA,
                              sd = NA,
                              se = NA,
                              ci = NA,
                              sum =  NA))

Fig_1B <- ggplot() +
  geom_sf(data = df, aes(color = sst), alpha = 0.7, size = 0.1) +
  geom_sf(data = world, fill = "gray30", color = "gray10") +
  theme_classic(base_size = 20) +
  theme(legend.position = "none") +
  theme(axis.title = element_blank()) +
  scale_color_viridis_c() + 
  scale_alpha_continuous(range = c(0,0.4)) +
  geom_point(data = centroids,
             aes(x = longitude, y = latitude, 
                 size = sum), color = "white", shape = 1) +
  geom_point(data = centroids,
             aes(x = longitude, y = latitude), size = 3, color = "black") +
  geom_point(data = centroids,
             aes(x = longitude, y = latitude), size = 2, color = "gold") +
  geom_label_repel(data = centroids, 
             aes(label = species, x = longitude, 
                 y = latitude), size = 4, nudge_y = 20, 
             segment.color = "gold") +
  scale_size_continuous(range = c(3.5,15))

# Export Fig 1B as png
png("~/Documents/GitHub/Long_AFP/Figures/Fig_1B.png", units = "in", width = 12, 
    height = 6, 
    res = 600)

Fig_1B
                   
```

Calculate freezing risk and plot against AFP copy number

```{r}

df_species_sf <- st_as_sf(centroids, coords = c("longitude", "latitude"), crs = 4326)

nearest_indices <- st_nearest_feature(df_species_sf, df)

df_species_sf$nearest_temperature <- df$sst[nearest_indices]

as.data.frame(df_species_sf)[,c(1,13)]

## Regression of crude freezing risk and AFP number
busco_df <- read.csv(("Busco_results_genome_size.csv"))

library(scales)

ggplot(data = busco_df,
       aes(x = Mode_lat*Risk*(1/UsMid_depth), y = AFPs)) +
  geom_point() +
  scale_x_continuous(trans = pseudo_log_trans(0.001, 10)) +
  geom_smooth(method = "lm", se = TRUE) +
  theme_classic()

Fig_A <- ggplot(data = busco_df,
       aes(x = Mode_lat*Risk*(1/UsMid_depth), y = AFPs)) +
  geom_smooth(method = "loess", se = TRUE, span = .95) +
  geom_point(size = 2) +
  scale_x_continuous(trans = pseudo_log_trans(0.01, 10)) +
  scale_y_continuous(trans = pseudo_log_trans(0.01, 10), breaks = c(0,1,2,5,10,25,50, 100, 1000, 1000)) +
  theme_classic() +
  labs(y = "log(AFP copy number)", x = "log(Polar x Latitude x 1/Depth)", title = "With depth")

Fig_B <- ggplot(data = busco_df,
       aes(x = Mode_lat*Risk, y = AFPs)) +
  #geom_smooth(method = "lm", se = TRUE) +
  geom_point(size = 2) +
  #scale_x_continuous(trans = pseudo_log_trans(0.1, 10)) +
  #scale_y_continuous(trans = pseudo_log_trans(0.1, 10), breaks = c(0,1,2,5,10,25,50, 100, 1000, 1000)) +
  theme_classic() +
  labs(y = "AFP copy number", x = "Polar x Latitude", title = "No depth")

# Arrange
Fig_1 <- ggarrange(Fig_A, Fig_B,
                   labels = c("A", "B"),
                   widths = c(1, 1),
                   ncol = 2, nrow = 1, align = "hv")

Fig_1

# No phylo correction
busco_df$fr_risk <- busco_df$Mode_lat * busco_df$Risk * (1/busco_df$UsMid_depth)
busco_df$pslog_risk <- log10(busco_df$fr_risk + (0.5*0.03236466))
busco_df$pslog_AFPs <- log10(busco_df$AFPs + (0.5*1))

busco_df$exp_risk <- busco_df$fr_risk^2


ggplot(busco_df,
       aes(x = pslog_risk, y = pslog_AFPs)) +
  geom_point() +
  geom_smooth(method = "lm", 
              se = TRUE)

# ANOVA without phylogenetically independent contrast
anova(lm(pslog_AFPs ~ pslog_risk, data = busco_df))

```


```{r}

# Read tree file
tree <- read.newick("Trees/Supermatrix_IQtree_WAGroot/SUPERMATRIX.phylip.treefile")

tree$tip.label <- gsub("busco_assembly_", "", tree$tip.label)

tree <- reroot(tree, 
             node.number = 12, 
             position = .5*(max(tree$edge.length)))

tree$tip.label <- c("Anarhichas minor", "Anarhichas lupus", "Lycodes diapterus", 
                    "Ophthalmolycus amberensis", "Lycodes pacificus", "Melanostigma gelatinosum",
                    "Lycodes platyrhinus", "Leptoclinus maculatus", "Pholis gunnellus", 
                    "Cebidichthys violaceus", "Bathymaster signatus", "Gasterosteus aculeatus")

# Plot AFP copy number
busco_df$tip.label = factor(
  busco_df$tip.label, 
  levels = rev(c('Ophthalmolycus amberensis', 'Lycodes diapterus', 'Lycodes pacificus', 
             'Lycodes platyrhinus', 'Melanostigma gelatinosum', 'Anarhichas lupus',
             'Anarhichas minor', 'Pholis gunnellus', 'Leptoclinus maculatus',
             'Cebidichthys violaceus', 'Bathymaster signatus', 'Gasterosteus aculeatus')))

p1 <- ggplot(busco_df, aes(y = tip.label, x = AFPs)) + 
  geom_bar(stat = "identity") +
  geom_text(aes(label=AFPs, y = tip.label), nudge_x = 5) +
  theme_tree2() + 
  theme(legend.position='none') +
  scale_x_continuous(limits = c(0,50)) +
  labs(title = "AFP copies")

library(cowplot)

```

```{r}

# Plot species tree

tree$node.label <- seq(1,11,1)

hl_nodes <- data.frame(id = c(10))

selected_nodes <- data.frame(nodes = c(5,4))

Fig_1A <- ggtree(tree, layout = "rectangular", size = 1) %<+% busco_df +
  geom_tiplab(color = "black", nudge_x = .005, align = TRUE, size = 5) +
  geom_tippoint(aes(size = Kelley_Lab), color = "skyblue3") +
  #geom_nodepoint(data = selected_nodes, aes(node = nodes), shape = "square") +
  theme(legend.position = "none",
        strip.background = element_blank(),
        axis.line = element_blank(),
        axis.ticks = element_blank()) +
  scale_size_manual(values = c(0,4)) +
  xlim(0, .175)

Fig_1A


```


```{r}
library(patchwork)

Fig_1Ap <- Fig_1A + p1 + plot_layout(widths = c(77, 23))

# Export Fig 1B as png
png("~/Documents/GitHub/Long_AFP/Figures/Fig_1A.png", units = "in", width = 8, 
    height = 6, 
    res = 600)

Fig_1A

```

```{r}

Fig_1 <- ggarrange(Fig_1Ap, Fig_1B,
                   labels = c("A", "B"),
                   widths = c(1, 1.5),
                   ncol = 2, nrow = 1, align = "hv")

Fig_1

# Export Fig 1A as png
png("~/Documents/GitHub/Long_AFP/Figures//Fig_1.png", units = "in", width = 24, 
    height = 6, 
    res = 600)

Fig_1

```

# AFPIII Gene Tree

```{r}

# Read in IQTree gene tree
IQ_genetree <- read.tree("Trees/Gene_Trees/01042024_IQTREE_K3Pu+F+G4_SAS_AFP_genetree.newick")

# Create factor for AFPs and SAS genes
group_df <- data.frame(tip.label = IQ_genetree$tip.label,
  group = ifelse(
  grepl("SAS", IQ_genetree$tip.label), "SAS",
  ifelse(grepl("AFP", IQ_genetree$tip.label), "AFP", NA)))

IQ_genetree <- reroot(IQ_genetree, 
             node.number = 193, 
             position = .5*(max(tree$edge.length)))

# Plot tree
ggtree(IQ_genetree, aes(color = group), layout = "rectangular") %<+% group_df +
  geom_tiplab(size = 1.5, align = TRUE) +
  geom_treescale() +
  labs(title = "Rooted IQtree gene tree K3Pu+F+G4 substitution model")


```

Phylogenetically independent contrast of AFP CNV

```{r}

# With phylo-independent contrast
pr_busco_df <- busco_df[-c(seq(13, 22, 1)), ]

# Transform to phy-independent values (PICs)
pic.pslogrisk <- pic(pr_busco_df$pslog_risk, tree)
pic.pslogAFPs <- pic(pr_busco_df$pslog_AFPs, tree)
pic.fr_risk <- pic(pr_busco_df$fr_risk, tree)
pic.AFPs <- pic(pr_busco_df$AFPs, tree)

# Correlation of PICs
cor.test(pic.pslogrisk, pic.pslogAFPs)
anova(lm(pic.pslogAFPs ~ pic.pslogrisk))
anova(lm(pic.AFPs ~ pic.fr_risk))


# Bayesian PIC
library(brms)

# Create dataframe of PIC values
brm_df <- data.frame(PIC_pslogrisk = pic.pslogrisk,
                     PIC_frrisk = pic.fr_risk,
                     PIC_pslog_AFPs = pic.pslogAFPs,
                     PIC_AFPS = pic.AFPs)

brm_df$sc_pslogrisk <- scale(brm_df$PIC_pslogrisk)
brm_df$sc_pslogAFPs <- scale(brm_df$PIC_pslog_AFPs)
brm_df$sc_risk <- scale(brm_df$PIC_frrisk)
brm_df$sc_AFPs <- scale(brm_df$PIC_AFPS)

# Fit brm
brm_AFP_frrisk <- brm(
  sc_pslogAFPs ~ sc_pslogrisk,
  data = brm_df,
  family = gaussian(),
  chains = 4, iter = 40000,
  save_pars = save_pars(TRUE)
)

summary(brm_AFP_frrisk)
plot(brm_AFP_frrisk)
conditional_effects(brm_AFP_frrisk)
pp_check(brm_AFP_frrisk)
bayes_R2(brm_AFP_frrisk)

ggplot(data = brm_df,
       aes(x = PIC_AFPS, y = PIC_pslog_AFPs)) +
  geom_point(size = 3, alpha = 0.5) +
  geom_smooth(method = "lm", span = 5, se = TRUE)



```

