---
title: "Reverse_Amplicon_ID"
author: "Sam Bogan"
date: "2023-11-09"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)

library(tidyverse)

```

```{r}

# Polar Genome Hits-Finding Matches
dfA_data <- read.csv()
dfB_data <- read.csv("Polar Genome Hits-Finding Matches.csv")

## Edit so that QS selection is replaced with selecting the minimum PF-PR distance from a group term for each primer

filter_and_select_max_qs <- function(df_A, df_B) {
  # Merge df_A and df_B on PrimerID and Chromosome to filter df_B
  merged_df <- merge(df_A, df_B, by = c("PrimerID", "Chromosome"))
  
  # Group by PrimerID and Chromosome and select the row with the max QS value
  result_df <- merged_df %>%
    group_by(PrimerID, Chromosome) %>%
    slice(which.max(QS))
  
  return(result_df)
}

# Example DataFrames
df_A <- data.frame(PrimerID = c('p1', 'p2'), Chromosome = c('chr1', 'chr2'))
df_B <- data.frame(
  PrimerID = c('p1', 'p2', 'p1', 'p2', 'p3'),
  Chromosome = c('chr1', 'chr2', 'chr1', 'chr2', 'chr2'),
  QS = c(0.1, 0.2, 0.3, 0.4, 0.5)
  )

# Run the function with the example data
filtered_df <- filter_and_select_max_qs(df_A, df_B)

print(filtered_df)

```

